File: domain/half_move.py
from domain.point import Point
from domain.color import Color

class HalfMove:
    def __init__(self, from_point: Point, to_point: Point, color: Color) -> None:
        self.from_point = from_point
        self.to_point = to_point
        self.color = color

    def __str__(self):
        return str(self.from_point.position) + "->" + str(self.to_point.position)

    def __repr__(self):
        return self.__str__()

    def is_valid(self) -> bool:
        return self.from_point.is_color(self.color) and self.to_point.is_open(self.color)

    def two_checkers_available(self) -> bool:
        return self.from_point.is_double_color(self.color)

    def merge(self, other: 'HalfMove'):
        return HalfMove(self.from_point, other.to_point, self.color)

    def can_merge(self, other: 'HalfMove'):
        return self.to_point == other.from_point

    def can_merge_or_vice_versa(self, other: 'HalfMove'):
        return self.can_merge(other) or other.can_merge(self)

File: ios/TavliEngine/Sources/TavliEngine/HalfMove.swift
import Foundation

/// Mirrors `domain/half_move.py`. Holds references to the board's `Point`
/// objects, so applying a half-move mutates the board in place.
public final class HalfMove: CustomStringConvertible, Equatable {
    public let from: Point
    public let to: Point
    public let color: Color

    public init(from: Point, to: Point, color: Color) {
        self.from = from
        self.to = to
        self.color = color
    }

    public var description: String { "\(from.position)->\(to.position)" }

    /// From-point is owned by color and to-point is landable.
    public func isValid() -> Bool {
        from.isColor(color) && to.isOpen(for: color)
    }

    public func twoCheckersAvailable() -> Bool { from.isDoubleColor(color) }

    public func canMerge(with other: HalfMove) -> Bool { to == other.from }

    public func canMergeOrViceVersa(with other: HalfMove) -> Bool {
        canMerge(with: other) || other.canMerge(with: self)
    }

    public func merge(with other: HalfMove) -> HalfMove {
        HalfMove(from: from, to: other.to, color: color)
    }

    public static func == (lhs: HalfMove, rhs: HalfMove) -> Bool {
        lhs.from == rhs.from && lhs.to == rhs.to && lhs.color == rhs.color
    }
}
